### PR TITLE
Adding PrefixConstrainedLogitsProcessor

### PR DESCRIPTION
**Pull request by nicola-decao**
# What does this PR do?
This pull request adds a new decoding strategy that constrains the next token to generate based on a callable function. It is a new PR that fixes https://github.com/huggingface/transformers/pull/7784 since the generate function went trough refactoring. @patrickvonplaten

### CODE REVIEW

**Comment by nicola-decao** [comment]
@patrickvonplaten I think we are ready to merge now 😊

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -366,6 +372,12 @@ def generate(
             use_cache: (:obj:`bool`, `optional`, defaults to :obj:`True`):
                 Whether or not the model should use the past last key/values attentions (if applicable to the model) to
                 speed up decoding.
+            prefix_allowed_tokens_fn: (:obj:`Callable`, `optional`, defaults to :obj:`None`):
+                If provided, it has to be a function that has as arguments :obj:`inputs_id`. At each step of Beam
+                Search, this function is called with the :obj:`inputs_id` containing the previously generated tokens as
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:377]: How many arguments does the function take? 2 -> `batch_id` and `inputs_id` or just `batch_id`?

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -366,6 +372,12 @@ def generate(
...
+                a tensor of shape :obj:`(batch_size * num_beams)`:. This function has to return a list of lists with
+                the allowed BPE tokens at the next step (list of batches and list of beams). This argument is useful
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:379]: maybe not say "BPE" tokens but just tokens => Otherwise people might think it only works with BPE.

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -300,6 +305,7 @@ def generate(
         num_return_sequences: Optional[int] = None,
         decoder_start_token_id: Optional[int] = None,
         use_cache: Optional[bool] = None,
+        prefix_allowed_tokens_fn: Callable = None,
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:308]: We should improve the type hinting here, either:

`Callable[[int, torch.Tensor], List[List[int]]` or `Callable[[torch.Tensor], List[List[int]]` if I understand correctly.

File: src/transformers/generation_logits_process.py [hunk 4]
```diff
@@ -372,3 +373,25 @@ def _set_scores_to_inf_for_banned_tokens(self, scores: torch.Tensor, banned_toke
...
         return scores
+
+
+class PrefixConstrainedLogitsProcessor(LogitsProcessor):
+    r"""
+    :class:`transformers.LogitsProcessor` .
+
+    Args:
+        prefix_allowed_tokens_fn (:obj:`Callable`):
+
```
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:384]: Docstring seems to be missing

File: src/transformers/generation_logits_process.py [hunk 5]
```diff
@@ -372,3 +373,25 @@ def _set_scores_to_inf_for_banned_tokens(self, scores: torch.Tensor, banned_toke
...
+    """
+
+    def __init__(self, prefix_allowed_tokens_fn: Callable, num_beams: int):
+        self._prefix_allowed_tokens_fn = prefix_allowed_tokens_fn
+        self._num_beams = num_beams
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+        mask = torch.full_like(scores, -math.inf)
+        for batch_id, beam_sent in enumerate(input_ids.view(-1, self._num_beams, input_ids.shape[-1])):
```
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:393]: In a future PR we could probably speed this up by just using `torch.Tensor` operations and not Python loops. Python loops really slow down the computation on GPU apparently (see: https://github.com/huggingface/transformers/pull/6064). But we can do this in a future PR as well

**Comment by patrickvonplaten** [comment]
Hey @nicola-decao,

Thanks a lot for re-implementing your PR! A couple of things I think we should improve.

1) I'm still not 100% sure how the API of `prefix_allowed_tokens_fn` would look like. Does it take one or two arguments? Also it would be very useful to correctly define the type hints in this case as I mentioned in the comment above

2) Can we add a test for this logits processor? It should be relatively straight-forward in /home/patrick/hugging_face/transformers/tests/test_generation_logits_process.py. The test should only test the `prefix_allowed_tokens_fn` not the whole generate function. 

3) It would be awesome if the docstring could be a bit more explicit (*e.g.* including your paper and maybe even a tiny example)

4) Can we delete the `src/.DS_Store ` file?

File: src/transformers/generation_logits_process.py (hunk 5 above)
Comment (nicola-decao) [reply src/transformers/generation_logits_process.py:396]: I wanted to keep the same signature as in fairseq as if someone has already implemented one it can use the same ;)

**Review by patrickvonplaten — APPROVED** [review]
Great looks good to me now! @yjernite - do you want to take a final look as well?

**Comment by yjernite** [comment]
> Great looks good to me now! @yjernite - do you want to take a final look as well?

Yup! Will take a look by EOD

**Comment by yjernite** [comment]
LGTM, thanks for implementing this functionality!

**Comment by patrickvonplaten** [comment]
Great work @nicola-decao

**Comment by nicola-decao** [comment]
> Great work @nicola-decao

Thanks :)
